## GPU/TPU 사용 설정

- Menu > Runtime > Change runtime type > Hardware accelerator > GPU 

In [0]:
## colab machine 확인
!cat /etc/issue

Ubuntu 18.04.3 LTS \n \l



In [0]:
## Memory 
!cat /proc/meminfo

MemTotal:       13335192 kB
MemFree:        10667016 kB
MemAvailable:   12493912 kB
Buffers:           70424 kB
Cached:          1910312 kB
SwapCached:            0 kB
Active:           690720 kB
Inactive:        1702684 kB
Active(anon):     388492 kB
Inactive(anon):      324 kB
Active(file):     302228 kB
Inactive(file):  1702360 kB
Unevictable:           0 kB
Mlocked:               0 kB
SwapTotal:             0 kB
SwapFree:              0 kB
Dirty:               276 kB
Writeback:             0 kB
AnonPages:        412736 kB
Mapped:           218468 kB
Shmem:               892 kB
Slab:             161064 kB
SReclaimable:     126608 kB
SUnreclaim:        34456 kB
KernelStack:        3712 kB
PageTables:         5528 kB
NFS_Unstable:          0 kB
Bounce:                0 kB
WritebackTmp:          0 kB
CommitLimit:     6667596 kB
Committed_AS:    2789304 kB
VmallocTotal:   34359738367 kB
VmallocUsed:           0 kB
VmallocChunk:          0 kB
AnonHugePages:         0 kB
ShmemHugePages:  

In [0]:
## CPU
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 85
model name	: Intel(R) Xeon(R) CPU @ 2.00GHz
stepping	: 3
microcode	: 0x1
cpu MHz		: 2000.172
cache size	: 39424 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs
bogomips	: 400

In [0]:
# GPU
!nvidia-smi

Tue Nov 19 01:05:27 2019       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 430.50       Driver Version: 418.67       CUDA Version: 10.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla P100-PCIE...  Off  | 00000000:00:04.0 Off |                    0 |
| N/A   32C    P0    26W / 250W |      0MiB / 16280MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU  

## Google Drive와 Colaboratory 연동

- Google Drive : pre-trained BERT model, KorQuAD 
- Colaboratory : Source code  

In [2]:
## 아래 코드를 실행하고 생성된 링크로 이동한 후 'Allow' 클릭, 인증코드 복사하여 붙여 넣기
from google.colab import drive
drive.mount('/content/gdrive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/gdrive


In [0]:
!ls /content/gdrive/My\ Drive

 ai-summit  'Colab Notebooks'   etc
 Classroom  'Epson iPrint'      thatsokye@raonbit.com


#### BERT multilingual model

[BERT](https://github.com/google-research/bert/blob/master/multilingual.md)

- BERT-Base, Multilingual Cased (New, recommended): 104 languages, 12-layer, 768-hidden, 12-heads, 110M parameters

In [0]:
## BERT 
import urllib.request
import zipfile

SAVE_DIR = '/content/gdrive/My Drive/ai-summit/model/'

def download(url, savedir=SAVE_DIR, zipped=True):
  file_name = url.split('/')[-1]
  download_file = savedir + file_name
  urllib.request.urlretrieve(url, download_file)

  if zipped:
    zip = zipfile.ZipFile(download_file)
    zip.extractall(savedir)
    zip.close()

url = 'https://storage.googleapis.com/bert_models/2018_11_23/multi_cased_L-12_H-768_A-12.zip'
download(url)

In [0]:
!ls /content/gdrive/My\ Drive/ai-summit/model/multi_cased_L-12_H-768_A-12

bert_config.json		     bert_model.ckpt.index  vocab.txt
bert_model.ckpt.data-00000-of-00001  bert_model.ckpt.meta


In [0]:
## vocabulary 확인

vocabfile = open(SAVE_DIR + 'multi_cased_L-12_H-768_A-12/vocab.txt', 'r')
vocab = vocabfile.readlines()

len(vocab)


119547

In [0]:
for v in vocab:
  if ord('가') <= ord(v[0]) <= ord('힣'):
    print(v)

vocabfile.close()

가

각

간

갇

갈

감

갑

값

갓

갔

강

갖

같

갚

개

객

갤

갯

갱

걀

거

걱

건

걷

걸

검

겁

것

겉

게

겐

겔

겟

겠

겨

격

겪

견

결

겸

겹

겼

경

곁

계

고

곡

곤

곧

골

곰

곱

곳

공

곶

과

곽

관

괄

괌

광

괜

괴

굉

교

구

국

군

굳

굴

굵

굽

굿

궁

권

궐

궤

귀

귄

규

균

그

극

근

글

금

급

긋

긍

기

긴

길

김

깁

깃

깊

까

깎

깐

깔

깜

깝

깥

깨

꺼

꺾

껍

껏

께

껴

꼈

꼬

꼭

꼴

꼼

꼽

꽃

꽤

꾀

꾸

꾼

꿀

꿈

꿔

꿨

뀌

뀐

끄

끈

끊

끌

끓

끔

끗

끝

끼

낀

낄

낌

나

낙

난

날

낡

남

납

낫

났

낭

낮

낯

낱

낳

내

낸

낼

냄

냈

냉

냐

냥

너

넌

널

넓

넘

넛

넣

네

넥

넨

넬

넷

녀

녁

년

념

녔

녕

노

녹

논

놀

놈

농

높

놓

뇌

뇨

누

눅

눈

눌

눠

뉘

뉜

뉴

늄

느

늑

는

늘

늙

능

늦

늬

니

닉

닌

닐

님

닛

닝

다

닥

닦

단

닫

달

닭

닮

담

답

닷

당

닿

대

댄

댈

댐

댓

더

덕

던

덜

덟

덤

덧

덩

덮

데

덱

덴

델

도

독

돈

돋

돌

돔

돕

돗

동

돼

됐

되

된

될

됨

두

둑

둔

둘

둠

둥

뒀

뒤

뒷

듀

듈

드

득

든

듣

들

듬

듭

듯

등

디

딕

딘

딜

딥

딧

딩

딪

따

딱

딴

딸

땃

땄

땅

때

떠

떡

떤

떨

떻

떼

또

똑

뚜

뚝

뚫

뚱

뛰

뜨

뜬

뜻

띄

띠

띤

라

락

란

랄

람

랍

랐

랑

래

랙

랜

램

랩

랫

랬

랭

랴

략

량

러

In [0]:
import json

configfile = open(SAVE_DIR + 'multi_cased_L-12_H-768_A-12/bert_config.json', 'r')
config = json.load(configfile)
config

{'attention_probs_dropout_prob': 0.1,
 'directionality': 'bidi',
 'hidden_act': 'gelu',
 'hidden_dropout_prob': 0.1,
 'hidden_size': 768,
 'initializer_range': 0.02,
 'intermediate_size': 3072,
 'max_position_embeddings': 512,
 'num_attention_heads': 12,
 'num_hidden_layers': 12,
 'pooler_fc_size': 768,
 'pooler_num_attention_heads': 12,
 'pooler_num_fc_layers': 3,
 'pooler_size_per_head': 128,
 'pooler_type': 'first_token_transform',
 'type_vocab_size': 2,
 'vocab_size': 119547}

In [0]:
configfile.close()

#### BERT Config
- attention_probs_dropout_prob': attention의 dropout 비율
- directionality': 'bidi',
- hidden_act': encoder와 pooler layer를 위한 비선형 activation function
- hidden_dropout_prob': 모든 FC(embeddings, encoder, pooler) layer의 dropout 비율
- hidden_size': encoder layers 와 pooler layer 의 크기
- initializer_range': 0.02,
- intermediate_size': Transformer encoder의 intermediate(feed-forward) layer의 크기
- max_position_embeddings': 이 모델에서 사용할 최대 문장 길이
- num_attention_heads': Transformer encoder안의 각 attention layer의 attention head의 개수
- num_hidden_layers': Transformer encoder의 hidden layer 개수
- pooler_fc_size': 768,
- pooler_num_attention_heads': 12,
- pooler_num_fc_layers': 3,
- pooler_size_per_head': 128,
- pooler_type': 'first_token_transform',
- type_vocab_size': BertModel에 전달되는 token_type_ids의 vocabulary 크기
- vocab_size': 119547

#### KorQuAD : 한국어 질문 답변 Dataset

[KorQuAD](https://korquad.github.io/)

- KorQuAD 2.1의 전체 데이터는 47,957 개의 Wikipedia article에 대해 102,960 개의 질의응답 쌍
- Training set 83,486 개, Dev set 10,165 개의 질의응답쌍으로 구분
- KorQuAD 2.0 데이터 중 HTML 태그의 속성이 완벽하게 제거되지 않은 오류를 수정

In [0]:
## KorQuAD

train = 'https://github.com/korquad/korquad.github.io/tree/master/dataset/KorQuAD_2.1/train'
dev = 'https://github.com/korquad/korquad.github.io/tree/master/dataset/KorQuAD_2.1/dev'


In [0]:
from bs4 import BeautifulSoup as bs

def listFiles(url, ext='zip'):
    page = requests.get(url).text
    soup = bs(page, 'html.parser')
    return [(url + '/' + node.get('href').split('/')[-1]).replace('tree', 'raw') for node in soup.find_all('a') if node.get('href').endswith(ext)]

train_files = listFiles(train)
train_files

['https://github.com/korquad/korquad.github.io/raw/master/dataset/KorQuAD_2.1/train/KorQuAD_2.1_train_00.zip',
 'https://github.com/korquad/korquad.github.io/raw/master/dataset/KorQuAD_2.1/train/KorQuAD_2.1_train_01.zip',
 'https://github.com/korquad/korquad.github.io/raw/master/dataset/KorQuAD_2.1/train/KorQuAD_2.1_train_02.zip',
 'https://github.com/korquad/korquad.github.io/raw/master/dataset/KorQuAD_2.1/train/KorQuAD_2.1_train_03.zip',
 'https://github.com/korquad/korquad.github.io/raw/master/dataset/KorQuAD_2.1/train/KorQuAD_2.1_train_04.zip',
 'https://github.com/korquad/korquad.github.io/raw/master/dataset/KorQuAD_2.1/train/KorQuAD_2.1_train_05.zip',
 'https://github.com/korquad/korquad.github.io/raw/master/dataset/KorQuAD_2.1/train/KorQuAD_2.1_train_06.zip',
 'https://github.com/korquad/korquad.github.io/raw/master/dataset/KorQuAD_2.1/train/KorQuAD_2.1_train_07.zip',
 'https://github.com/korquad/korquad.github.io/raw/master/dataset/KorQuAD_2.1/train/KorQuAD_2.1_train_08.zip',
 

In [0]:
dev_files = listFiles(dev)

for t in train_files:
  print(t)
  download(t, savedir=SAVE_DIR+'train/')

for d in dev_files:
  print(d)
  download(d, savedir=SAVE_DIR+'dev/')

https://github.com/korquad/korquad.github.io/raw/master/dataset/KorQuAD_2.1/train/KorQuAD_2.1_train_00.zip
https://github.com/korquad/korquad.github.io/raw/master/dataset/KorQuAD_2.1/train/KorQuAD_2.1_train_01.zip
https://github.com/korquad/korquad.github.io/raw/master/dataset/KorQuAD_2.1/train/KorQuAD_2.1_train_02.zip
https://github.com/korquad/korquad.github.io/raw/master/dataset/KorQuAD_2.1/train/KorQuAD_2.1_train_03.zip
https://github.com/korquad/korquad.github.io/raw/master/dataset/KorQuAD_2.1/train/KorQuAD_2.1_train_04.zip
https://github.com/korquad/korquad.github.io/raw/master/dataset/KorQuAD_2.1/train/KorQuAD_2.1_train_05.zip
https://github.com/korquad/korquad.github.io/raw/master/dataset/KorQuAD_2.1/train/KorQuAD_2.1_train_06.zip
https://github.com/korquad/korquad.github.io/raw/master/dataset/KorQuAD_2.1/train/KorQuAD_2.1_train_07.zip
https://github.com/korquad/korquad.github.io/raw/master/dataset/KorQuAD_2.1/train/KorQuAD_2.1_train_08.zip
https://github.com/korquad/korquad.gi

In [0]:
!ls /content/gdrive/My\ Drive/ai-summit/model/train

korquad2.1_train_00.json  korquad2.1_train_09.json  korquad2.1_train_23.json
KorQuAD_2.1_train_00.zip  KorQuAD_2.1_train_09.zip  korquad2.1_train_24.json
korquad2.1_train_01.json  korquad2.1_train_10.json  korquad2.1_train_25.json
KorQuAD_2.1_train_01.zip  KorQuAD_2.1_train_10.zip  korquad2.1_train_26.json
korquad2.1_train_02.json  korquad2.1_train_11.json  korquad2.1_train_27.json
KorQuAD_2.1_train_02.zip  KorQuAD_2.1_train_11.zip  korquad2.1_train_28.json
korquad2.1_train_03.json  korquad2.1_train_12.json  korquad2.1_train_29.json
KorQuAD_2.1_train_03.zip  KorQuAD_2.1_train_12.zip  korquad2.1_train_30.json
korquad2.1_train_04.json  korquad2.1_train_13.json  korquad2.1_train_31.json
KorQuAD_2.1_train_04.zip  korquad2.1_train_14.json  korquad2.1_train_32.json
korquad2.1_train_05.json  korquad2.1_train_15.json  korquad2.1_train_33.json
KorQuAD_2.1_train_05.zip  korquad2.1_train_16.json  korquad2.1_train_34.json
korquad2.1_train_06.json  korquad2.1_train_17.json  korquad2.1_train_35.json

In [0]:
!rm -rf /content/gdrive/My\ Drive/ai-summit/model/*.zip
!rm -rf /content/gdrive/My\ Drive/ai-summit/model/train/*.zip
!rm -rf /content/gdrive/My\ Drive/ai-summit/model/dev/*.zip

In [0]:
!ls /content/gdrive/My\ Drive/ai-summit/model/dev

korquad2.1_dev_00.json	korquad2.1_dev_02.json	korquad2.1_dev_04.json
korquad2.1_dev_01.json	korquad2.1_dev_03.json


In [0]:
import json

def showJson(f, showcnt=0):
  json_file = open(f)
  json_data = json.load(json_file)
  cnt = len(json_data['data'])
  if showcnt >= cnt:
    showcnt = con
  print(json.dumps(json_data['data'][showcnt], indent=2))

In [14]:
showJson('/content/gdrive/My Drive/ai-summit/model/train/korquad2.1_train_00.json', 20)

{
  "title": "\ud0dc\ud48d_\uc140\ub9c8_(1987\ub144)",
  "url": "https://ko.wikipedia.org/wiki/\ud0dc\ud48d_\uc140\ub9c8_(1987\ub144)",
  "context": "<!DOCTYPE html>\n<html>\n<head>\n<meta>\n<title>\ud0dc\ud48d \uc140\ub9c8 (1987\ub144) - \uc704\ud0a4\ubc31\uacfc, \uc6b0\ub9ac \ubaa8\ub450\uc758 \ubc31\uacfc\uc0ac\uc804</title>\n\n\n<link>\n\n<meta>\n<link>\n<meta>\n<meta>\n<meta>\n<meta>\n<meta>\n<link>\n<link>\n<link>\n<link>\n<link>\n<link>\n<link>\n<link>\n<link>\n<link>\n<link>\n\n</head>\n<body>\n<div></div>\n<div></div>\n<div>\n<a></a>\n<div></div>\n<div>\n</div>\n<h1>\ud0dc\ud48d \uc140\ub9c8 (1987\ub144)</h1>\n<div>\n<div>\uc704\ud0a4\ubc31\uacfc, \uc6b0\ub9ac \ubaa8\ub450\uc758 \ubc31\uacfc\uc0ac\uc804.</div>\n<div></div>\n<div></div>\n<a>\ub458\ub7ec\ubcf4\uae30\ub85c \uac00\uae30</a>\n<a>\uac80\uc0c9\ud558\ub7ec \uac00\uae30</a>\n<div><div><table>\n<caption><b>198705 \uc140\ub9c8 (THELMA)</b>\n</caption>\n<tbody><tr>\n<th colspan=\"3\"><b>\ucd08\ub300\ud615\uc758</b> <b>\ub

In [20]:
## 본문
json_data['data'][20]['context']

'<!DOCTYPE html>\n<html>\n<head>\n<meta>\n<title>태풍 셀마 (1987년) - 위키백과, 우리 모두의 백과사전</title>\n\n\n<link>\n\n<meta>\n<link>\n<meta>\n<meta>\n<meta>\n<meta>\n<meta>\n<link>\n<link>\n<link>\n<link>\n<link>\n<link>\n<link>\n<link>\n<link>\n<link>\n<link>\n\n</head>\n<body>\n<div></div>\n<div></div>\n<div>\n<a></a>\n<div></div>\n<div>\n</div>\n<h1>태풍 셀마 (1987년)</h1>\n<div>\n<div>위키백과, 우리 모두의 백과사전.</div>\n<div></div>\n<div></div>\n<a>둘러보기로 가기</a>\n<a>검색하러 가기</a>\n<div><div><table>\n<caption><b>198705 셀마 (THELMA)</b>\n</caption>\n<tbody><tr>\n<th colspan="3"><b>초대형의</b> <b>매우 강한 태풍</b> <b>(<a>JMA</a>)</b> <a><img></a>\n</th></tr>\n<tr>\n<th colspan="3"><b>4등급의 슈퍼 태풍</b> <b>(<a>SSHS</a>)</b>\n</th></tr>\n<tr>\n<td colspan="3"><a><img></a><br/>태풍 셀마, 최성기의 모습 (7월 11일)\n</td></tr>\n<tr>\n<td><b>발생일</b>\n</td>\n<td colspan="2"><a>1987년</a> <a>7월 9일</a>\n</td></tr>\n<tr>\n<td><b>소멸일</b>\n</td>\n<td colspan="2"><a>1987년</a> <a>7월 16일</a>\n</td></tr>\n<tr>\n<td><b>최저 기압</b>\n</td>\n<td colspan="2">911h

In [21]:
## 질문
json_data['data'][20]['qas'][0]['question']

'87년 태풍 셸마의 근원이 되는 열대저기압은 언제 발생하였을까?'

In [17]:
## 답변
json_data['data'][20]['qas'][0]['answer']['text']

'<a>7월 7일</a> 오전 9시'